### DISCUSSION TITLE
Edge conditions with non-Lagrange elements

### DISCUSSION THREAD

**Discussion by pzuritas** [firedrakeproject/firedrake#2821] [Firedrake support] [answered]
Hello! I am trying to implement a mixed Poisson problem in Firedrake and my installation raises the following error when trying to define Dirichlet boundary conditions:

`TypeError: Currently, edge conditions have only been tested with Lagrange elements`

The error is simple enough to understand, but I'm curious as to why this is the case when the [mixed Poisson tutorial](https://www.firedrakeproject.org/demos/poisson_mixed.py.html) assigns boundary conditions with Brezzi-Douglas-Marini elements. ¿Is the tutorial outdated, or am I missing something?

If necessary, please see the MRE below.

```Python
from firedrake import (
    BoxMesh, FunctionSpace, FacetNormal, TrialFunctions, TestFunctions, inner,
    div, dx, ds, DirichletBC, as_vector, Function, solve
)


mesh = BoxMesh(5, 5, 5, 10, 10, 10)
V_h = FunctionSpace(mesh, 'BDM', 1)
Q_h = FunctionSpace(mesh, 'DG', 0)
W_h = V_h * Q_h
n = FacetNormal(mesh)

(u, p) = TrialFunctions(W_h)
(v, q) = TestFunctions(W_h)

a = inner(u, v)*dx
a += inner(p, div(v))*dx
a += inner(div(u), q)*dx

L = inner(v, n)*ds(1)

bc_in = DirichletBC(W_h.sub(0), as_vector([1.0, 0.0, 0.0]), 1)
bc_air = DirichletBC(W_h.sub(0), as_vector([0.0, 0.0, 0.0]), [range(3, 7)])

x = Function(W_h)
solve(a == L, x, bcs=[bc_in, bc_air])

# Error is raised, but conflicts with
# https://www.firedrakeproject.org/demos/poisson_mixed.py.html
```

**Comment by ksagiyam** [comment, marked as answer]
I think what you meant to do is:
```
bc_air = DirichletBC(W_h.sub(0), as_vector([0.0, 0.0, 0.0]), [*range(3, 7)])  # or [3, 4, 5, 6]
```
instead of:
```
bc_air = DirichletBC(W_h.sub(0), as_vector([0.0, 0.0, 0.0]), [range(3, 7)])
```

The most general form of `sub_domain` argument is tuple of tuples; if `sub_domain = ((3, 4), (5, ), (4, 5, 6))`, for instance, the boundary condition is applied on the intersection of boundaries 3 and 4, on boundary 5, and on the intersection of boundaries 4, 5, and 6. Currently, this most general form is only useful for Lagrange elements. (I should document this.)
If you pass a tuple `(3, 4, 5, 6)`, say, as `sub_domain`, it is internally converted to `((3, ), (4, ), (5, ), (6, ))`.
`[range(3, 7)]` is equivalent in this scope to `((3, 4, 5, 6), )`, so Firedrake tries to apply the boundary condition on the intersection of boundaries 3, 4, 5, and 6, which is not what you want here.

**Reply by pzuritas** [reply]
Great! That seemed to be it, and it makes sense then that the message read "edge" and not "boundary". Thanks so much for the comprehensive answer!
